packages/asu-core/src/auth/services/cas.service.ts

```diff
@@ -0,0 +1,77 @@
+import {singleton} from "tsyringe";
+import {AxiosProvider} from "../../http/providers/axios.provider";
+import {Axios} from "axios";
+import {APIGatewayProxyEvent, APIGatewayProxyResult} from "aws-lambda";
+import {UnauthenticatedError} from "../../errors/unauthenticated.error";
+import {lambdaRedirect} from "../../aws/lambda/response-handler";
+import {lambdaHandleError} from "../../aws/lambda/error-handler";
+import {EnvironmentVariable} from "../../enums/environment-variable";
+
+const CAS_HOST = 'https://weblogin.asu.edu/cas';
+
+@singleton()
+export class CasService {
+  private httpClient: Axios;
+  private authUser: string;
+
+  constructor(private axiosProvider: AxiosProvider) {
+  }
+
+  private async getHttpClient(): Promise<Axios> {
+    if (!this.httpClient) {
+      this.httpClient = this.axiosProvider.resolve({
+        baseURL: CAS_HOST
+      });
+    }
+
+    return this.httpClient;
+  }
+
+  public async authenticate(event: APIGatewayProxyEvent, callback: ()=>Promise<APIGatewayProxyResult>): Promise<APIGatewayProxyResult> {
+    // Bypass auth on dev
+    if (process.env[EnvironmentVariable.Stage] === 'dev') {
+      if (event.queryStringParameters && event.queryStringParameters.skip_auth === 'y') {
+        this.authUser = 'skipped_auth';
+        return await callback();
+      }
+    }
+
+    try {
+      //@todo get app host from env
+      const serviceUrl = 'https://localhost.asu.edu' + event.path;
+      if (!event.queryStringParameters || !event.queryStringParameters.ticket) {
+        return lambdaRedirect(`${CAS_HOST}/login?service=${serviceUrl}`);
+      }
+
+      const httpClient = await this.getHttpClient();
+      const authResponse = await httpClient.get('/serviceValidate', {
+        params: {
+          format: 'JSON',
+          service: serviceUrl,
+          ticket: event.queryStringParameters.ticket
+        }
+      });
+
+      const {authenticationSuccess, authenticationFailure} = authResponse.data.serviceResponse;
+
+      if (authenticationSuccess) {
+        this.authUser = authenticationSuccess.user;
+        return await callback();
+      }
+
+      console.log(`CasService.authenticate error: ${authenticationFailure.code} - ${authenticationFailure.description}`);
+      return lambdaHandleError(new UnauthenticatedError('Authentication request failed'));
```

Comment:
If the auth request returns a 200, but is an `authenticationFailure`, do you think we should just redirect the user back to CAS to try again? My apprehension with this is that the user might end up in an infinite auth loop if the error persists. 

Here are the potential error codes from CAS.serviceValidate endpoint: https://apereo.github.io/cas/6.5.x/protocol/CAS-Protocol-Specification.html#253-error-codes

My thought is that anything aside from `INVALID_TICKET` would be an internal error on our part and should not be retried. And even in the case of `INVALID_TICKET`, I find it very unlikely that the ticket will have somehow expired in such a short window since CAS will have just created it before redirecting back to our service.

So for now, just went with returning an error response.

Reply:
I was struggling with the same question earlier this week. I am not sure what we will want to do, but I think this pattern is totally fine for now.